src/core/thread/network_data_publisher.cpp

```diff
@@ -0,0 +1,1089 @@
+
+exit:
+    return;
+}
+
+void Publisher::PrefixEntry::CountOnMeshPrefixEntries(uint8_t &aNumEntries, uint8_t &aNumPreferredEntries) const
+{
+    const PrefixTlv *      prefixTlv;
+    const BorderRouterTlv *brSubTlv;
+    int8_t                 preference             = BorderRouterEntry::PreferenceFromFlags(mFlags);
+    uint16_t               flagsWithoutPreference = BorderRouterEntry::FlagsWithoutPreference(mFlags);
+
+    prefixTlv = Get<Leader>().FindPrefix(mPrefix);
+    VerifyOrExit(prefixTlv != nullptr);
+
+    brSubTlv = prefixTlv->FindSubTlv<BorderRouterTlv>(/* aStable */ true);
+    VerifyOrExit(brSubTlv != nullptr);
+
+    for (const BorderRouterEntry *entry = brSubTlv->GetFirstEntry(); entry <= brSubTlv->GetLastEntry();
+         entry                          = entry->GetNext())
+    {
+        uint16_t entryFlags      = entry->GetFlags();
+        int8_t   entryPreference = BorderRouterEntry::PreferenceFromFlags(entryFlags);
+
+        // Count an existing entry in the network data if its flags
+        // match ours and and its preference is same or higher than our
+        // preference. We do not count matching entries at a lower
+        // preference than ours. This ensures that a device with higher
+        // preference entry publishes its entry even when there are many
+        // lower preference similar entries in the network data
+        // (potentially causing a lower preference entry to be removed).
+
+        if ((BorderRouterEntry::FlagsWithoutPreference(entryFlags) == flagsWithoutPreference) &&
+            (entryPreference >= preference))
+        {
+            aNumEntries++;
+
+            // We prefer an entry if it has strictly higher preference
+            // than ours or if it has same preference we use the associated
+            // RLOC16.
+
+            if ((entryPreference > preference) || IsPreferred(entry->GetRloc()))
+            {
+                aNumPreferredEntries++;
+            }
+        }
+    }
+
+exit:
+    return;
+}
+
+void Publisher::PrefixEntry::CountExternalRouteEntries(uint8_t &aNumEntries, uint8_t &aNumPreferredEntries) const
+{
+    const PrefixTlv *  prefixTlv;
+    const HasRouteTlv *hrSubTlv;
+    int8_t             preference             = HasRouteEntry::PreferenceFromFlags(static_cast<uint8_t>(mFlags));
+    uint8_t            flagsWithoutPreference = HasRouteEntry::FlagsWithoutPreference(static_cast<uint8_t>(mFlags));
+
+    prefixTlv = Get<Leader>().FindPrefix(mPrefix);
+    VerifyOrExit(prefixTlv != nullptr);
+
+    hrSubTlv = prefixTlv->FindSubTlv<HasRouteTlv>(/* aStable */ true);
+    VerifyOrExit(hrSubTlv != nullptr);
+
+    for (const HasRouteEntry *entry = hrSubTlv->GetFirstEntry(); entry <= hrSubTlv->GetLastEntry();
+         entry                      = entry->GetNext())
+    {
+        uint8_t entryFlags      = entry->GetFlags();
+        int8_t  entryPreference = BorderRouterEntry::PreferenceFromFlags(entryFlags);
```

Comment:
```suggestion
        int8_t  entryPreference = HasRouteEntry::PreferenceFromFlags(entryFlags);
```

Reply:
Thanks for noticing. Fixed this in new push.

I was wondering why the test did not catch this, turns out this caused the `entryPreference` to become zero (which is same "med" level which we used in the test. I changed the test to use "low" instead.